### PR TITLE
[modules] Accept equivalent module caches from different symlink

### PR DESCRIPTION
**Pull request by ellishg**
Use `VFS.equivalent()`, which follows symlinks, to check if two module cache paths are equivalent. This prevents a PCH error when building from a different path that is a symlink of the original.

```
error: PCH was compiled with module cache path '/home/foo/blah/ModuleCache/2IBP1TNT8OR8D', but the path is currently '/data/users/foo/blah/ModuleCache/2IBP1TNT8OR8D'
1 error generated.
```

### CODE REVIEW

File: clang/lib/Serialization/ASTReader.cpp [hunk 1]
```diff
@@ -839,17 +839,15 @@ static bool checkHeaderSearchOptions(const HeaderSearchOptions &HSOpts,
...
-        Diags->Report(diag::err_pch_modulecache_mismatch)
-          << SpecificModuleCachePath << ExistingModuleCachePath;
-      return true;
-    }
-  }
-
-  return false;
+  if (!LangOpts.Modules || PPOpts.AllowPCHWithDifferentModulesCachePath ||
+      SpecificModuleCachePath == ExistingModuleCachePath ||
+      llvm::sys::fs::equivalent(SpecificModuleCachePath,
```
Comment (benlangmuir) [inline clang/lib/Serialization/ASTReader.cpp:844]: This should go through the `VirtualFileSystem` to be consistent with the code that reads module files. The callers of this API should both have easy access to a VFS to pass in: for `SimplePCHValidator` it's `FileMgr.getVirtualFileSystem()` and for `PCHValidator` it's `Reader.getFileManager().getVirtualFileSystem()`.

You could also add a helper method to VirtualFileSystem for `equivalent` that does `status(Path1).equivalent(status(Path2))` (plus error handling).

Comment (ellishg) [reply clang/lib/Serialization/ASTReader.cpp]: Thanks for the advice. I realized that this does not handle the case where both paths are equal, but do not exist. In that case, no error will be thrown. Is that acceptable? I guess this was the case before this pr.

Comment (benlangmuir) [reply clang/lib/Serialization/ASTReader.cpp]: That's fine, it will fail when it tries to read anything from the cache.

**Comment by ellishg** [comment]
@benlangmuir do you have other concerns? Can I get a stamp?

File: clang/lib/Serialization/ASTReader.cpp [hunk 2]
```diff
@@ -833,32 +833,33 @@ bool SimpleASTReaderListener::ReadPreprocessorOptions(
 /// against the header search options in an existing preprocessor.
 ///
 /// \param Diags If non-null, produce diagnostics for any mismatches incurred.
-static bool checkHeaderSearchOptions(const HeaderSearchOptions &HSOpts,
+static bool checkHeaderSearchOptions(llvm::vfs::FileSystem &VFS,
```
Comment (benlangmuir) [inline clang/lib/Serialization/ASTReader.cpp:836]: I realize you're not changing the _behaviour_ here, but now that we don't even pass in a `HeaderSearchOptions` we should probably rename this. Maybe `checkModuleCachePath`?

**Review by benlangmuir — APPROVED** [review]
LGTM if we rename the checkHeaderSearchOptions function.

**Pull request merged**